### ISSUE TITLE
Warn on function call in `unwrap_or(..)` parameter

### ISSUE THREAD

**Issue by mqudsi** [rust-lang/rust#54041]
I believe the `Result::unwrap_or(..)` has a high potential for causing problems or unexpected behavior to users without previous intimate familiarity with lower-level languages, as it can cause unexpected side effects to take place.

For example, it is completely plausible that someone could write the following and be surprised by the output:

```rust
fn get_alternative() -> i32 {
    eprintln!("Forced to use alternative approach!");
    return 42;
}

fn main() {
    let r: Result<_, !> = Ok(7);
    println!("{}", r.unwrap_or(get_alternative()));
}
```

Where `get_alternative()` could have even more serious side effects than an incorrect error message. 

While `unwrap_or_else(|...| ...)` exists, I believe this is not sufficient to prevent this class of error from happening. The documentation on the reasons for the existence of `unwrap_or_else` vs `unwrap_or` (apart their names being complete artifacts due to rust's strict no overloading policy, only further adding to the confusion) does not stress the importance of this aspect of their differences.

Furthermore, given the sheer abundance of `Option` and `Result` return values in the Rust standard library, this has the potential to significantly optimize runtimes if people are doing something like this:

```rust
some_value.map(|x| x.foo.unwrap_or(fallback())).unwrap_or(fallback)
```

Where `fallback` is called twice to initialize just one value, when it doesn't even need to be called _at all_ in the normal case.

I can't think of a good story for _wanting_ the nested function calls in a `unwrap_or(...)` call to be executed regardless of whether the unwrap succeeded or failed, and if someone really wants it, the inconvenience of having them silence the warning with a #`[allow(unwrap_or_fn_call)]` is, imho, not a big deal compared to the entire class of errors implementing this warning would provide.

 (If it were up to me, I would have originally translated both `unwrap_or` and `unwrap_or_else` to a `match` statement in pre-compilation and forced delayed execution of nested function calls.)

**Proposal**

Compiling code containing something along the lines of the following:
```rust
    let foo = bar.unwrap_or(get_alternative());
```

Would generate ~an error~ a warning message:

> Nested function calls in Result::unwrap_or expressions are executed unconditionally. Did you mean to use Result::unwrap_or_else(...) instead?

**Comment by euclio** [comment]
There is already a [clippy lint](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#or_fun_call) for this exact scenario.

**Comment by Mark-Simulacrum** [comment]
We have a policy of accepting lints only through RFCs, and, either way, there would be false positives. In some cases I'd imagine it is more friendly to the optimizer to know that both paths will be executed or the function is cheap enough that the call itself might be more expensive.

**Comment by mqudsi** [comment]
@Mark-Simulacrum thanks for the response. Is there a list of what needs to be submitted as an RFC?

**Comment by Mark-Simulacrum** [comment]
Not that I know of - mostly anything that requires community feedback or needs discussion, I think.
